File: 17220.java
import java.util.*;
import java.io.*;

class Main{
    static int n,m;
    static ArrayList<Integer>[] graph;
    static List<Integer> header = new ArrayList<>();
    static int[] indegree;
    static boolean[] check;
    static int cnt;
    
    public static void main(String[] args) throws Exception{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        indegree = new int[n];
        check = new boolean[n];
        Arrays.fill(indegree,0);

        graph = new ArrayList[n];
        for(int i = 0; i < n; i++) graph[i] = new ArrayList<>();
        
        for(int i = 0; i < m; i++){
            st = new StringTokenizer(br.readLine());
            int from = String.valueOf(st.nextToken()).charAt(0) - 'A';
            int to = String.valueOf(st.nextToken()).charAt(0) - 'A';
            
            graph[from].add(to);
            indegree[to]++;
        }

        for(int i = 0; i < n; i++){
            if(indegree[i] == 0){
                header.add(i);
            }
        }

        st = new StringTokenizer(br.readLine());
        int m = Integer.parseInt(st.nextToken());
        for(int i = 0; i < m; i++){
            int a = String.valueOf(st.nextToken()).charAt(0) - 'A';

            graph[a].clear();
            if(!header.contains(a)){
                for(int j = 0; j < n; j++){
                    graph[j].remove(Integer.valueOf(a));
                }
            }
        }

        for(int i = 0; i < n; i++){
            if(indegree[i] == 0){
                dfs(i);               
            }
        }
        
        System.out.println(cnt);
    }

    static void dfs(int idx){
        check[idx] = true;
        for(int i = 0; i < graph[idx].size(); i++){
            if(!check[graph[idx].get(i)]){
                dfs(graph[idx].get(i));
                cnt++;
            }
        }
    }
}

File: 17220.cpp
#include <iostream>
#include <map>
#include <vector>

using namespace std;

int n,m,ans;
vector<int> graph[26];
vector<int> header;
// map<char,int> m;
int outDegree[26];
bool visited[26];

void dfs(int idx){
    for(int i = 0; i < graph[idx].size(); i++){
        int next = graph[idx][i];
        if(visited[next]) continue;
        ans++;
        dfs(next);
    }
}

int main(){
    cin >> n >> m;
    
    for(int i = 0; i < m; i++){
        char to,from;
        cin >> from >> to;
        graph[from-'A'].push_back((to-'A'));
        outDegree[to-'A']++;
    }

    for(int i = 0; i < n; i++){
        if(outDegree[i] == 0) header.push_back(i);
    }

    int k;
    cin >> k;
    for(int i = 0; i < k; i++){
        char inp;
        cin >> inp;
        visited[inp-'A'] = true;
    }

    for(int i = 0; i < header.size(); i++){
        int root = header[i];
        if(!visited[root]){
            visited[root] = true;
            dfs(root);
        }
    }

    cout << ans << "\n";
}